src/jit/flowgraph.cpp

```diff
@@ -1621,15 +1694,152 @@ void Compiler::fgReplacePred(BasicBlock* block, BasicBlock* oldPred, BasicBlock*
+    // and change this to an assert.
+    //
+    // Tracked with https://github.com/dotnet/coreclr/issues/13345
+    if (!movingPred)
+        return;
+
+    // Splice out the node that we are modifying.
+    if (*elementBeforeMovingPred)
+    {
+        *elementBeforeMovingPred = movingPred->flNext;
+    }
+
+    // Set the block to the new pred.
+    movingPred->flBlock = newPred;
+
+    // If the list is empty or the item is to be inserted at the head.
+    if (block->bbPreds == nullptr)
+    {
+        // We are inserting the newPred at the end of the list.AddArgumentToTail
+        block->bbPreds     = movingPred;
+        movingPred->flNext = nullptr;
+
+        fgExpensiveSortedPredCheck(block);
+        return;
+    }
+
+    flowList *pred = block->bbPreds, *prevPred = block->bbPreds;
+    bool      inserted = false;
+
+    // If inserting at the head of the list.
+    if (pred->flBlock->bbNum > newPred->bbNum)
+    {
+        block->bbPreds     = movingPred;
+        movingPred->flNext = pred;
+
+        fgExpensiveSortedPredCheck(block);
+        return;
+    }
+
+    // Add the new pred.
+    do
+    {
+        // We need to insert the newPred before the next pred.
+        if (pred->flBlock->bbNum > newPred->bbNum)
+        {
+            // We are inserting a new pred that does not exist
+            // already in the list. We can just link it in.
+
+            prevPred->flNext   = movingPred;
+            movingPred->flNext = pred;
+            assert(pred->flNext != pred);
+
+            inserted = true;
             break;
         }
+
+        // Equal
+        else if (pred->flBlock->bbNum == newPred->bbNum)
+        {
+            // We need to update this pred with a potentially higher ref count
+            // and drop the moving pred instead of reinserting it
+            // back in the pred list.
+
+            unsigned largerCount = pred->flDupCount;
+            largerCount          = movingPred->flDupCount > largerCount ? movingPred->flDupCount : largerCount;
+
+            noway_assert(pred->flDupCount > 0);
+            pred->flDupCount = largerCount;
+
+            inserted = true;
+            break;
+        }
+
+        prevPred = pred;
+        pred     = pred->flNext;
+    } while (pred);
+
+    // If we have not already inserted then prevPred points to the last element
+    // in the list and we need to insert at the end.
+    if (!inserted)
+    {
+        prevPred->flNext   = movingPred;
+        movingPred->flNext = nullptr;
+    }
+
+#if DEBUG
+
+    if (JitConfig.JitSlowDebugChecksEnabled() != 0)
+    {
+        unsigned refCount = 0;
+
+        // Make sure the ref count to this block is still correct.
+        for (flowList* pred = block->bbPreds; pred; pred = pred->flNext)
+        {
+            refCount += pred->flDupCount;
+        }
+
+        // If refCounts were correct coming in, but modifying incorrectly
+        // going out assert.
+        if (refCountCorrect)
```

Comment:
Nit: You could add this check above and avoid doing this if we already know it wasn't correct coming in.